### DISCUSSION TITLE
Text align for Number input

### DISCUSSION THREAD

**Discussion by DMazykin** [zauberzeug/nicegui#1918] [Q&A] [answered]
### Question

How can I set correctly alignment of text (entered value) for Number input. 
I actually thought that it should be by default "right" for number input based on Quasar "type=number".

Currently it's by default "left" and even if I set class "text-right" it is still "left"

```py
from nicegui import ui

ui.number(label="Number").classes("text-right")

ui.run()
```

**Comment by falkoschindler** [comment, marked as answer]
Because `ui.input` ([QInput](https://quasar.dev/vue-components/input)) is converted into a rather complex hierarchy of HTML elements, `.classes()` doesn't affect the `<input>` tag but some other container. You can use the "input-class" or "input-style" props instead to address the input element:
```py
ui.number(label='Number').props('input-class="text-right"')
```

**Reply by DMazykin** [reply]
Thanks @falkoschindler

**Comment by UlrichGoebel** [comment]
That's great. Can it be done, that the Label is shown right-aligned too? I tried

`ui.input('ID').props('dense input-class="text-right" label-class="w-full text-right"')`

but it doesn't work.

**Reply by evnchn** [reply]
*Posted by Claude Code on @evnchn's behalf.*

**TL;DR:** Quasar has no `label-class`/`label-align` prop for QInput, so your attempt is silently ignored — but it *is* doable: one small CSS rule on `.q-field__label`. This is the intended Quasar way (they treat label styling as a CSS concern, not a prop).

```py
from nicegui import ui

ui.add_css('''
.right-label .q-field__label {
    width: 100%;
    text-align: right;
    transform-origin: top right;  /* keep the float-up animation anchored on the right */
}
''')

ui.input('ID').props('input-class="text-right"').classes('right-label')

ui.run()
```

Right-aligns the label in both the placeholder and floated states (verified on NiceGUI 3.13):

<img width="600" alt="Before/after: label-class ignored (label left) vs CSS on .q-field__label (label right)" src="https://github.com/user-attachments/assets/0fb8a136-5d19-4884-be96-181c5a56e008" />


<details>
<summary><b>Why <code>label-class</code> did nothing</b></summary>

Unlike `input-class`/`input-style`, **QInput/QField has no `label-class` prop**, so `props('label-class="…"')` is dropped. (`labelClass` *is* a real prop in Quasar — but only on QFab/QInnerLoading/QKnob; in QField the label's class is a hardcoded internal constant `"q-field__label no-pointer-events absolute ellipsis"`.) Confirmed by grepping the bundled `quasar.umd.js` — `props.labelClass` is never read in the QField label render.

</details>

<details>
<summary><b>Refuted: the native <code>label</code> slot can't right-align it either</b></summary>

QField does expose a `label` slot (`add_slot('label')`), so this looks like the native route — but it can't do the alignment. Tested three variants (slot alone / slot + `label` prop / slot + `stack-label`); all stayed left:

| Approach | Label result |
|---|---|
| `label-class="w-full text-right"` (your attempt) | left — prop ignored |
| `add_slot('label')` + `full-width text-right` content | **left** |
| `add_slot('label')` + `stack-label` | **left** |
| CSS on `.q-field__label` (width + text-align) | **right** ✓ |

**Root cause:** the blocker isn't the label *content*, it's the container. Quasar renders `.q-field__label` as `position:absolute; left:0` and **shrink-wrapped to its text** (~15px wide). Anything *inside* it (slot content, utility classes) only has those ~15px to align within. The only thing that reaches the container's width is CSS on `.q-field__label` itself — hence the rule above. `transform-origin: top right` makes it scale toward the right corner as it floats, staying aligned with the value.

</details>

<details>
<summary><b>This is the Quasar-sanctioned approach, not a hack</b></summary>

Quasar deliberately has no prop for this — label/input styling is treated as a CSS concern:

- [quasarframework/quasar#2963](https://github.com/quasarframework/quasar/issues/2963) — maintainer on input text-align: *"this is not a problem with quasar but with devland CSS"* (target the inner element).
- [#2049 "[wish] label-align"](https://github.com/quasarframework/quasar/issues/2049) — a native `label-align` was requested, but only for the **old pre-v1** horizontal QField; closed "no longer relevant in v1." Never existed in modern Quasar.
- [#7116](https://github.com/quasarframework/quasar/issues/7116) / [discussion #13587](https://github.com/quasarframework/quasar/discussions/13587) — community label-positioning answers all route through CSS on `.q-field__label`.

`input-class` exists only because right-aligning input *text* was common enough to warrant a shortcut; the label never got one.

</details>

---
<sub>Changelog — 2026-06-23: added before/after screenshot.</sub>

**Reply by falkoschindler** [reply]
Or if you want to inline the custom styling using Tailwind's arbitrary-variant child-selector syntax:
```py
ui.input('ID').props('input-class="text-right"') \
    .classes(r'[&_.q-field\_\_label]:w-full [&_.q-field\_\_label]:text-right [&_.q-field\_\_label]:origin-top-right')
```

**Reply by UlrichGoebel** [reply]
Thank's a lot! I prefer the `ui.add_css`-Variant, but that's of course my personal preference.

**Reply by evnchn** [reply]
I also like the `ui.add_css` variant. 

I think we should not move to "Tailwind Shenanigans" all the time, because it just makes the code more coupled.

**Reply by falkoschindler** [reply]
Not all the time. I just meant to mention an alternative.
Sometimes I prefer to have styling tied closely to the element instead of adding page-wide CSS.
